src/Utilities/DoubleDicts.jl

```diff
@@ -3,116 +3,106 @@ module DoubleDicts
+
+Works as a `AbstractDict{MOI.ConstraintIndex,V}` with minimal differences.
 
-When optimal performance or type stability is required it is possible to obtain a
-fully type stable dictionary with values of type `V` and keys of type
-`CI{MOI.SingleVariable, MOI.Integers}` from the dictionary `dict`, for instance:
+If `V` is also a `MOI.ConstraintIndex`, use [`IndexDoubleDict`](@ref).
 
-    ``inner = dict[MOI.SingleVariable, MOI.Integers]``
+Note that `MOI.ConstraintIndex` is not a concrete type, opposed to
+`MOI.ConstraintIndex{MOI.SingleVariable, MOI.Integers}`, which is a concrete
+type.
+
+When looping through multiple keys of the same Function-in-Set type, use
+```julia
+inner = dict[F, S]
+```
+to return a type-stable [`DoubleDictInner`](@ref).
 """
-struct DoubleDict{V,DI,DO} <: AbstractDoubleDict{DataTypePair,V,Int64,DI,DO}
-    dict::DO
+struct DoubleDict{V} <: AbstractDoubleDict{V}
+    dict::Dict{Tuple{DataType,DataType},Dict{Int64,V}}
     function DoubleDict{V}() where {V}
-        return new{V,Dict{Int64,V},Dict{DataTypePair,Dict{Int64,V}}}(
-            Dict{DataTypePair,Dict{Int64,V}}(),
-        )
+        return new{V}(Dict{Tuple{DataType,DataType},Dict{Int64,V}}())
     end
 end
 
 """
-    IndexDoubleDict
+    DoubleDictInner{F,S,V}
+
+A type stable inner dictionary of [`DoubleDict`](@ref).
+"""
+mutable struct DoubleDictInner{F,S,V} <: AbstractDoubleDictInner{F,S,V}
+    dict::DoubleDict{V}
+    inner::Dict{Int64,V}
+    function DoubleDictInner{F,S}(d::DoubleDict{V}) where {F,S,V}
+        return new{F,S,V}(d, get!(d.dict, (F, S), Dict{Int64,V}()))
+    end
+end
 
-Specialized version of `DoubleDict` in which keys and values are of type
-`ConstraintIndex` (`CI`).
+function Base.getindex(d::DoubleDict, ::Type{F}, ::Type{S}) where {F,S}
+    return DoubleDictInner{F,S}(d)
+end
 
-This is an optimized dictionary to map `ConstraintIndex` (`CI`) to values of type `CI`.
-Works as a `AbstractDict{CI, V}` with minimal differences.
-Note that `CI` is not a concrete type, opposed to `CI{MOI.SingleVariable, MOI.Integers}`,
-which is a concrete type.
+"""
+    IndexDoubleDict
 
-When optimal performance or type stability is required its possible to obtain a
-fully type stable dictionary with values of type
-`CI{MOI.SingleVariable, MOI.Integers}` and keys of type
-`CI{MOI.SingleVariable, MOI.Integers}` from the dictionary `dict`, for instance:
+A specialized version of [`DoubleDict`] in which the values are of type
+`MOI.ConstraintIndex`
 
-    ``inner = dict[MOI.SingleVariable, MOI.Integers]``
+When looping through multiple keys of the same Function-in-Set type, use
+```julia
+inner = dict[F, S]
+```
+to return a type-stable [`IndexDoubleDictInner`](@ref).
 """
-struct IndexDoubleDict{DI,DO} <: AbstractDoubleDict{DataTypePair,CI,Int64,DI,DO}
-    dict::DO
+struct IndexDoubleDict <: AbstractDoubleDict{MOI.ConstraintIndex}
+    dict::Dict{Tuple{DataType,DataType},Dict{Int64,Int64}}
```

Comment:
It can be done in a separate PR but I think we should use a `CleverDict` here. With VectorOfConstraints and MatrixOfConstraints, the constraint indices are now 1:n for scalar constraints so using CleverDict here would speed up copy time.

Reply:
Yup, that is why it was previously parametrized by dict.
That is for the second Dict....
And.... we might consider LittleDict for the outer Dict.... (it is from OrderedCollections) leave it as comment?

Reply:
> so using CleverDict here would speed up copy time for large LP's

Benchmark? I struggle to believe that the time looking up the index translation is significant. I'm against adding complexity, even if it is a minor speed improvement.

Reply:
When doing the benchmarks for MatrixOfConstraints, I noticed creating the IndexMap was taking a significant time.

Reply:
In a proper model? Or in a microbenchmark?